### ISSUE TITLE
Primus support & Bumblebee 4.0

### ISSUE THREAD

**Issue by ArchangeGabriel** [Bumblebee-Project/Bumblebee#319]
Some logs from our discussion about this: https://bumblebee.etherpad.mozilla.org/8

Work is on common-primus branch.

I will list the new features soon.

Work remaining:
- revert my last commit on the three
- fix issues in primus

Some questions:
1) Based on this https://github.com/Bumblebee-Project/Bumblebee/commit/7fb2a8250aeca5fae805c185152fdf3c10875d2f, this https://github.com/Bumblebee-Project/bumblebee-ppa/blob/primus-proposed/precise/primus/debian/patches/ubuntu_libgl_path.patch and this https://github.com/amonakov/primus/blob/master/primusrun, how should we package primus ? Sould we hardcode some default path at compile time like it is done right now, plus eventually some thing in the primus script, or move everything to daemon side (I think that even the primusrun script is useless in this situation, since everything is set by the daemon, that can even make the final call, right) ? I personaly prefer the second option.

2) How is currently managed the fact that primus only start X on a GL call ?

**Comment by ArchangeGabriel** [comment]
I’ve started cleaning and sorting again the issue tracker, hope to finish that today. Before releasing, I would like my above xorg.conf change to be tested (will ask for that on the french forum), since we don’t seem to be in a fast release emergency anymore.

Also, I would like to update documentation, I’m going to use this thread to add note about things to do.

To be added in documentation:

```
ACPI Warning: \_SB_.PCI0.PEG0.PEGP._DSM: Argument #4 type mismatch - Found [Buffer], ACPI requires [Package] (20140724/nsarguments-95)
```
```
ACPI Error: Field [TMPB] at 282624 exceeds Buffer [ROM1] size 262144 (bits) (20140724/dsopcode-236)
ACPI Error: Method parse/execution failed [\_SB_.PCI0.PEG0.PEGP._ROM] (Node ffff88041f07db40), AE_AML_BUFFER_LIMIT (20140724/psparse-536)
```

**Comment by ArchangeGabriel** [comment]
I’ve been looking at setting `"AllowEmptyInitialConfiguration"` by default but it makes NVIDIA to detect the CRT/DFP link when one exist again, and I remember that with switched fully to `"UseDisplayDevice" "none"` because it makes the secondary X server to start a bit faster and configuring display in this case is useless. So we should indeed let this as it is currently and add documentation on online support about this.

I’m still waiting for more returns on the xorg.conf ignore device change, but didn’t get any negative ones as of today.

**Comment by ArchangeGabriel** [comment]
OK, so except for documentation updates, the only thing I’m not sure about is nvidia-uvm handling. I need to recheck what we planned to do and verify it will work everywhere.

**Comment by ArchangeGabriel** [comment]
We have a bit more issues to look into before releasing (all tagged with Milestone 4.0). I will open a pull request with the xorg.conf snippet and ask user to try it. I’m opening an issue about nvidia modules handling.

**Comment by ArchangeGabriel** [comment]
Do we want to keep the bumblebee-bugreport tool? We need at least to check which part of it are still relevant.

What about glsanity? Should we suggest distros to package it separately? Or do we just link it in bug reporting docs?

**Comment by Lekensteyn** [comment]
bumblebee-bugreport has much Debian/Ubuntu-specific things like update-alternatives, but the library path configuration (broken or not) is still universa (maybe glsanity checks that too, I have not looked at it).

With modern machines there is an issue with bbswitch that I am trying to address in https://github.com/Bumblebee-Project/bbswitch/tree/acpi-pr3 (basically, Windows 8 and newer decided to disable/enable power resources on the parent PCIe device, so support for the `_DSM` calls might become unreliable). A fix for that that is forward compatible with kernel 4.7 (in development) and newer requires that bbswitch acts as a PCI driver such that it can use runtime PM, but this approach needs some changes to bumblebeed for reliablity:

 - If the PMMethod is bbswitch and the loaded driver is "bbswitch", then assume it is turned off (and do not try to unload the bbswitch module!)
 - Instead of using the `/proc/acpi/bbswitch` interface, use the unbind/bind method described in https://github.com/Bumblebee-Project/bbswitch/commit/daa6411911426246aef804ec5ad8a44a07a15a66. This part is something I am not sure of. The problem with using the /proc/acpi/bbswitch interface is that it cannot really probe one specific device. What can be done is to trigger the probe for a device, but if other drivers are also available, they might bind with the device instead of bbswitch. With the unbind/new_id/unbind approach (only stable as userspace interface), the outcome is more deterministic. Hmm, while bumblebeed is used, there should be no other driver loaded when bbswitch is asked to turn it OFF, so maybe this potential issue can be ignored.

Do you think it is worth holding off BB 4.0 until bbswitch is fixed for the newer machines? Should solve issues like "memory corruption on Lenovos" and "battery dies quickly in suspend".

**Comment by ArchangeGabriel** [comment]
Yes, definitively worth. I thought at one point Bumblebee 4.0 should be a companion of bbswitch 1.0, but wasn’t sure where you were on that one. But since you’ve asked, yes, we should delay BB 4.0 until bb 1.0 is ready too. ;) Anyway, I still have a lot of documentation update to go for BB 4.0, so that should give you some time to get things done. :)

**Comment by bluca** [comment]
We backport the good new stuff in Debian and in the Ubuntu PPA anyway, so users have a way to get new bug fixes and features until this is finished too.

**Comment by ArchangeGabriel** [comment]
So, we decided to go without waiting bbswitch 1.0 finally. We still agree on removing nouveau support, right? This will need a point in the release note (like for people wanting to switch dynamically to nouveau, stop bumblebeed daemon and `modprobe -r bbswitch`, then `modprobe nouveau` and `DRI_PRIME=1 <program>` — which require DRI3, but I’ll suppose any system with bumblebee 4.0 will be DRI3 anyway — instead of optirun + instructions for switching back to bumblebee mode).

I’ll go through the open issues and task ASAP, and I propose a tentative release date as next sunday (30th October).

**Comment by Lekensteyn** [comment]
Yes, drop nouveau support from Bumblebee and write/point to some documentation on using nouveau properly in an Optimus setup (nouveau wiki?).

FYI, I will be less available in the next 2-3 weeks due to exams and delayed project work ;)

**Comment by ArchangeGabriel** [comment]
Sadly, I had (and still have) to postpone this on my side too, I’ve been overbusy+seek for the past 10 days, and I’ve just seen I won’t have any free time available before at least december…

**Comment by Lekensteyn** [comment]
Ok, no problem, maybe I can do some additional QA now that I can test the nvidia blob again. Still in exam period, but it is an idea for the moments thereafter :)
